### DISCUSSION TITLE
How to get text from a descendent of a descendent?

### DISCUSSION THREAD

**Discussion by Harsha-Gelato** [appium/appium#17558] [General]
<img width="440" alt="Screenshot 2022-10-04 at 11 11 40" src="https://user-images.githubusercontent.com/103107117/193818812-a7f9081a-d044-427b-91d2-21869352a807.png">

Hi Team,

I want to get text "**From" "£8.60"** as you can see in above heirachy of my flutter application. I am automating my flutter **application using Appium flutter driver in typeScript language.** 
I want to get list of child Text widgets and read text from them iteratively and then concatenate all the texts into one and check does it match the expected one.
What is the best way to achieve this?

I tried below but I cannot able to getElementText() like this. When my test run it just tries to get text but it freeze at that time because it fails to do so and timeout error.

```
const catalogPrice = find.descendant({
            of: find.byValueKey(`catalog_product_#${index}`),
            matching: find.byValueKey('catalog_product_card_price'),
        });
        const textWidget = find.descendant({
            of: find.byValueKey(catalogPrice),
            matching: find.byType('Text'),
            matchRoot: true,
            firstMatchOnly: false,
        });

        console.log('$$$$$$$$$$$$$$$$$$');
        console.log(textWidget);
        console.log('@@@@@@@@@@@@@@@@@@@@@@');
        const price = await driver.getElementText(textWidget);
```

It would be very helpful if someone can please let me know what I am missing at my end. Thanks :)

**Comment by Harsha-Gelato** [comment]
@KazuCocoa Could you please provide some input on this issue? It would be very helpful. Thanks :)

**Reply by KazuCocoa** [reply]
I do not have much idea, but 

> it freeze at that time because it fails to do so and timeout error.

then the Dart VM internal did not handle the query well...?
Does the element exist in the getRenderTree result?

**Reply by Harsha-Gelato** [reply]
Dart VM is handling the query well. Is below a correct way to find text from a descendent of a descendent?
```
const catalogPrice = find.descendant({
            of: find.byValueKey(`catalog_product_#${index}`),
            matching: find.byValueKey('catalog_product_card_price'),
        });
        const textWidget = find.descendant({
            of: find.byValueKey(catalogPrice),
            matching: find.byType('Text'),
            matchRoot: true,
            firstMatchOnly: false,
        });
```
I have got text previously from a descendent directly and it worked fine but now I have to get text from a descendent of a descendent, that where I have a doubt

**Reply by Harsha-Gelato** [reply]
And yes the element exists in the getRenderTree result, please find the screenshot as well attached in my post. If you look at the screenshot, in this tree, I want to get text present inside the wrap but those text does not have unique keys only Wrap is having a key name. In such situation what is the best way to get text eg: **"£8.60"**?

**Reply by Harsha-Gelato** [reply]
[render.txt](https://github.com/appium/appium/files/9714496/render.txt)
@KazuCocoa I am also attaching the render output for the same to take the reference. please let me know if anything else is required from my end.

**Comment by Mr-Ples** [comment]
If you first add `console.log(catalogPrice);` does it freeze there?
